blog: add "feature flags are just if statements, right?" post

agardnerIT opened the pull request:
## This PR
- Adds a blog post

beeme1mr commented on blog/27-03-2023-feature-flags-if-statements.md:
```diff
@@ -0,0 +1,62 @@
+- If user is premium and located in Australia = v4 (orange)
+- Everyone else gets v1 (green)
+That’s 2 variables. Imagine how many variations a real system has. Now imagine trying to update these conditions in real time, without a redeployment.
+<img src={require('@site/static/img/blog/feature-flags-if-statements/feature-flag-context.png').default} />
```
```suggestion

<img src={require('@site/static/img/blog/feature-flags-if-statements/feature-flag-context.png').default} />

```

beeme1mr reviewed (commented):
There seem to be some spacing issues. Can you please run this locally and add spacing after bullet points and headers?

agardnerIT commented:
Can I request a re-review on this please?

toddbaert commented on blog/27-03-2023-feature-flags-if-statements.md:
```diff
@@ -0,0 +1,64 @@
+draft: false
+---
+
+Most feature flag explainers begin by explaining that feature flags are glorified if statements.
```
I agree with the idea that this is the naive view, but I'm not sure most blog posts or explainers take that view. I would consider rephrasing this to something that suggests that `feature flags == env vars` is the default naive or primitive implementation.

toddbaert commented on blog/27-03-2023-feature-flags-if-statements.md:
````diff
@@ -0,0 +1,64 @@
+}
+```
+Now try modelling this in code:
+- If user is premium and located in Europe = v2 (purple)
+- If user is premium and located in the US = v3 (red)
+- If user is premium and located in Australia = v4 (orange)
+- Everyone else gets v1 (green)
````
It might be worth adding another example or two, like a psuedo-random assignment, perhaps even downstream of another rule.

toddbaert commented on blog/27-03-2023-feature-flags-if-statements.md:
```diff
@@ -0,0 +1,64 @@
+<img src={require('@site/static/img/blog/feature-flags-if-statements/feature-flag-context.png').default} />
+
+The real power of feature flags:
+- They can be as simple or as complex as you like (this makes getting started easy with plenty of room to grow and expand)
```
The first point seems to be a proper sentence, it has a subject ("They can be...") while the others are sentence fragments and don't include the subject "they". This makes the points feel a bit disjointed.

toddbaert commented on blog/27-03-2023-feature-flags-if-statements.md:
```diff
@@ -0,0 +1,64 @@
+- Separation of concerns: Allows you to deploy without releasing (place all new features behind flags and disable them).
+- Shortcut to continuous delivery: If you can deploy anything at any time with zero risk, why not do it?
+- Separation of responsibilities: Product managers take responsibility of feature flags, leaving DevOps teams to deal with their area of responsibility.
+## So, are Feature Flags if Statements?
```
```suggestion

## So, are Feature Flags if Statements?

```

toddbaert commented on blog/27-03-2023-feature-flags-if-statements.md:
```diff
@@ -0,0 +1,64 @@
+### 1: OpenFeature Demo
+Never encountered feature flags?
+This is the 30,000ft view showing the power and versatility of feature flags.
+[Start here](https://bit.ly/openfeaturedemo).
```
I agree.

If we're going to use a URL shortener, it should be with an account owned by the community as well.

toddbaert commented on blog/27-03-2023-feature-flags-if-statements.md:
```diff
@@ -0,0 +1,64 @@
+Choose a vendor or code your own?
+[Start your journey](https://bit.ly/openfeatureflagd) with the open source flagd solution.
+### 3. Instrument Your Application
+Whether you've chosen to use flagd, gone with a vendor solution, or developed a custom OpenFeature-compliant in-house solution, by now your backend flag evaluation infrastructure is running.
```
```suggestion
Whether you've chosen to use a community project, gone with a vendor solution, or developed a custom OpenFeature-compliant in-house solution, by now your backend flag evaluation infrastructure is running.
```

toddbaert commented on blog/27-03-2023-feature-flags-if-statements.md:
```diff
@@ -0,0 +1,64 @@
+[Start your journey](https://bit.ly/openfeatureflagd) with the open source flagd solution.
+### 3. Instrument Your Application
+Whether you've chosen to use flagd, gone with a vendor solution, or developed a custom OpenFeature-compliant in-house solution, by now your backend flag evaluation infrastructure is running.
+It is time to instrument your application!
+The [Five minutes to feature flags](https://bit.ly/openfeature5mins) tutorial walks through how to instrument an application in an OpenFeature-compliant way.
```
This is the first we've heard of instrumentation or telemetry. I think it might be worth mentioning earlier that one of the biggest benefits of OpenFeature is common extension and integration points, and standards which form a good basis for telemetry.

toddbaert commented on blog/27-03-2023-feature-flags-if-statements.md:
```diff
@@ -0,0 +1,64 @@
+Whether you've chosen to use flagd, gone with a vendor solution, or developed a custom OpenFeature-compliant in-house solution, by now your backend flag evaluation infrastructure is running.
+It is time to instrument your application!
+The [Five minutes to feature flags](https://bit.ly/openfeature5mins) tutorial walks through how to instrument an application in an OpenFeature-compliant way.
+### 4: Feature Flags on Kubernetes?
+Are you Kubernetes-native and want to manage and use feature flags entirely on Kubernetes?
+The [Operator tutorial](https://bit.ly/openfeatureopdemo) demonstrates how to leverage Custom Resources to manage feature flags via GitOps and make feature flags available to your pods via annotations.
```
I'd recommend removing this section. It again feels like it's pushing flagd too much, and the OpenFeature blog isn't the right place for that IMHO.

beeme1mr commented on blog/27-03-2023-feature-flags-if-statements.md:
````diff
@@ -0,0 +1,64 @@
+  return v1
+}
+```
+Now try modelling this in code:
````
I guess that's US vs UK spelling. In the past, we've gone with US but that requirement isn't explicitly stated anywhere.

agardnerIT commented:
I've addressed the feedback and made it more generic but I don't agree that removing references to flagd (entirely) is appropriate.

Flagd is an open source, OpenFeature compliant, reference implementation. As such, IMO, it is entirely appropriate to direct people to use it (if they aren't already using feature flags).

I entirely agree that we should never say "flagd is the way" but "flagd is one way" is entirely appropriate, especially for brand new users.
